### PR TITLE
DTLS 1.3: additions for event driven server in wolfssl-examples

### PR DESCRIPTION
**Pull request by julek-wolfssl**
- Add callback when we parse a verified ClientHello
- Allow the DTLS server to operate without maintaining state

### CODE REVIEW

**Review by dgarske — CHANGES_REQUESTED** [review]
What is driving the addition of this client_hello good callback? Did you consider expanding the `WOLFSSL_CALLBACKS` and `HandShakeCallBack` used by `wolfSSL_accept_ex` / `wolfSSL_connect_ex` to support this feature?
If adding an API and callback function would be nice to not limit it to client_hello. This callback seems very DTLS v1.3 specific.

File: src/tls.c [hunk 1]
```diff
@@ -6163,8 +6163,16 @@ static int TLSX_Cookie_Parse(WOLFSSL* ssl, const byte* input, word16 length,
 
     /* client_hello */
     extension = TLSX_Find(ssl->extensions, TLSX_COOKIE);
-    if (extension == NULL)
-        return HRR_COOKIE_ERROR;
+    if (extension == NULL) {
+#ifdef WOLFSSL_DTLS13
+        if (ssl->options.dtls)
```
Comment (rizlik) [inline src/tls.c:6168]: Is this for dtls 1.3 only? I think you should add `IsAtLeastv13` to the check

**Review by rizlik — CHANGES_REQUESTED** [review]
Is the CB supposed to work with DTLSv1.0 and v1.2 as well?

**Comment by julek-wolfssl** [comment]
> Is the CB supposed to work with DTLSv1.0 and v1.2 as well?

It should but I haven't tested it yet.

> What is driving the addition of this client_hello good callback?

The motivation behind the new callback is to signal to the user that we have performed a successful return-routability test with the current peer. This means that the current object should dedicated to only serving this connection and a new object should be created to listen for new connections. I will see if I can make use of the existing handshake callbacks instead of writing a new callback. I agree that the callback is very DTLS specific.

File: src/internal.c [hunk 2]
```diff
@@ -8783,6 +8783,14 @@ static int EdDSA_Update(WOLFSSL* ssl, const byte* data, int sz)
 int HashRaw(WOLFSSL* ssl, const byte* data, int sz)
 {
     int ret = 0;
+#ifdef WOLFSSL_DEBUG_TLS
+    byte digest[WC_MAX_DIGEST_SIZE];
+
+    WOLFSSL_MSG("HashRaw:");
+    WOLFSSL_MSG("Data:");
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/internal.c:8790]: The DEBUG_TLS logs are only useful for dev work already so I think it makes sense to include as much data as possible. This also helps to figure out where we went wrong with the hash.

File: src/tls.c [hunk 3]
```diff
@@ -6163,8 +6163,16 @@ static int TLSX_Cookie_Parse(WOLFSSL* ssl, const byte* input, word16 length,
...
+            /* TODO: Should we allow a ClientHello with a valid cookie even if
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/tls.c:6169]: I think we should because this might happen in multi-thread load balancing. A different thread replied earlier but now we got the CH2 so I don't think this should be an error.

File: src/tls.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/tls.c:6168]: The cookie extension is only present in TLS 1.3 (I think) so you are right. I will add it.

**Comment by julek-wolfssl** [comment]
@dgarske I took a look at the `WOLFSSL_CALLBACKS` handshake callbacks but I don't think that this is right solution here. These callbacks only inform the user what packets have been processed and the `hsCb` callback only gets called after the handshake finishes. There is also the `protoMsgCb` callback but that is gated behind `OPENSSL_EXTRA` and it still doesn't provide any information if the ClientHello is verified or not. This will not work for DTLS because we want to isolate the socket as soon as we have processed a verified ClientHello. For this I think that a dedicated callback for DTLS is reasonable.

File: src/ssl.c [hunk 4]
```diff
@@ -1183,6 +1183,14 @@ int wolfSSL_dtls_set_peer(WOLFSSL* ssl, void* peer, unsigned int peerSz)
     if (ssl == NULL)
         return WOLFSSL_FAILURE;
 
+    if (peer == NULL || peerSz == 0) {
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c]: Added docs.

File: src/ssl.c [hunk 5]
```diff
@@ -12745,6 +12765,20 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
 
 #endif /* NO_WOLFSSL_SERVER */
 
+#ifdef WOLFSSL_DTLS
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c]: Wrapped.

File: src/wolfio.c [hunk 6]
```diff
@@ -382,11 +382,35 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
...
+        if (dtlsCtx->peer.sa == NULL) {
+            dtlsCtx->peer.sa = (void*)XMALLOC(sizeof(SOCKADDR_S),
+                    ssl->heap, DYNAMIC_TYPE_SOCKADDR);
+            dtlsCtx->peer.sz = 0;
+            if (dtlsCtx->peer.sa != NULL)
+                dtlsCtx->peer.bufSz = sizeof(SOCKADDR_S);
+            else
+                dtlsCtx->peer.bufSz = 0;
+        }
+        peer = dtlsCtx->peer.sa;
```
Comment (dgarske) [inline src/wolfio.c:410]:
```
./configure '--enable-all' '--enable-sp-math-all' 'CC=g++' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK' && make
...
src/wolfio.c:410:30: error: invalid conversion from ‘void*’ to ‘SOCKADDR_S*’ {aka ‘sockaddr_storage*’} [-fpermissive]
  410 |         peer = dtlsCtx->peer.sa;
      |                ~~~~~~~~~~~~~~^~
      |                              |
      |                              void*
```

File: wolfssl/internal.h [hunk 7]
```diff
@@ -4460,6 +4463,11 @@ struct WOLFSSL {
 #ifdef WOLFSSL_STATIC_MEMORY
     WOLFSSL_HEAP_HINT heap_hint;
 #endif
+#ifdef WOLFSSL_DTLS
```
Comment (dgarske) [inline wolfssl/internal.h:4466]: Do we always need this DTLS feature? Perhaps we can add a new macro for it? Shouldn't it be disabled if building with `NO_WOLFSSL_SERVER`? I want to make sure embedded DTLS customers with small flash/ram don't have issues.

File: wolfssl/internal.h [hunk 8]
```diff
@@ -2201,13 +2201,16 @@ WOLFSSL_LOCAL int DoVerifyCallback(WOLFSSL_CERT_MANAGER* cm, WOLFSSL* ssl,
...
+    unsigned int bufSz; /* sockaddr size */
     void*        sa; /* pointer to the sockaddr_in or sockaddr_in6 */
 };
 
 typedef struct WOLFSSL_DTLS_CTX {
     WOLFSSL_SOCKADDR peer;
     int rfd;
     int wfd;
+    byte userSet:1;
+    byte connected:1; /* Set when the rfd and wfd are connected sockets */
```
Comment (dgarske) [inline wolfssl/internal.h:2213]: Can we come up with a better name for this member? It seems to only change the use of the "peer" sockaddr in the recvfrom and sendto.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfssl/internal.h (hunk 7 above)
Comment (julek-wolfssl) [reply wolfssl/internal.h:4466]: I'm not sure putting it behind a new define makes sense. Adding a `!NO_WOLFSSL_SERVER` is a good idea since this is only useful for a server.

File: src/wolfio.c (hunk 6 above)
Comment (julek-wolfssl) [reply src/wolfio.c]: Fixed cast.

File: wolfssl/ssl.h [hunk 9]
```diff
@@ -1045,6 +1045,7 @@ WOLFSSL_API int wolfSSL_CTX_set1_param(WOLFSSL_CTX* ctx, WOLFSSL_X509_VERIFY_PAR
 WOLFSSL_API int  wolfSSL_is_server(WOLFSSL* ssl);
 WOLFSSL_API WOLFSSL* wolfSSL_write_dup(WOLFSSL* ssl);
 WOLFSSL_ABI WOLFSSL_API int  wolfSSL_set_fd(WOLFSSL* ssl, int fd);
+WOLFSSL_API int wolfSSL_set_dtls_fd_connected(WOLFSSL* ssl, int fd);
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply wolfssl/ssl.h:1048]: Wrapped.

File: wolfssl/internal.h (hunk 8 above)
Comment (julek-wolfssl) [reply wolfssl/internal.h:2213]: I honestly can't think of a better name that won't turn into a Java-esque sentence. When `connected` is set, this indicates that the sockets in `rfd` and `wfd` are "connected" (`connect()` was called on them in addition to `bind`). This means that `sendto` and `recvfrom` don't need to specify and store the peer address. Also if I remember correctly, @ejohnstown reported on another PR that using `sendto` on a connected socket on a Mac resulted on a failure. This may require us to switch to using just `send` in `EmbedSendTo` on a connected socket.

In short: can you think of a better name to describe this behaviour because I can't?

**Review by ejohnstown — CHANGES_REQUESTED** [review]
When I tested this locally, merged with master (0459e83), the unit.test would fail:
```
starting DTLSv1.3 suite - downgrade
trying server command line[3571]: SuiteTest -vd -7 2 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -2 -p 0 
trying client command line[3572]: SuiteTest -v 3 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -f -2 -p 51930 
SSL_accept error -326, record layer version error
wolfSSL error: SSL_accept failed
```
I used the configuration:
```
john:~/src/wolfssl (julek-wolfssl-dtls-good-ch-cb) % ./config.status --config
--enable-dtls --enable-dtls13 --enable-hrrcookie
```

**Review by dgarske — APPROVED** [review]
Renamed the function to DTLS specific. Did not try and run the code. John did and he's having issues (see his comment). I will wait to merge until those are resolved.

**Comment by rizlik** [comment]
> When I tested this locally, merged with master ([0459e83](https://github.com/wolfSSL/wolfssl/commit/0459e83a59d83ac7ebba3271a09ec26c52a56642)), the unit.test would fail:
> 
> ```
> starting DTLSv1.3 suite - downgrade
> trying server command line[3571]: SuiteTest -vd -7 2 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -2 -p 0 
> trying client command line[3572]: SuiteTest -v 3 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -f -2 -p 51930 
> SSL_accept error -326, record layer version error
> wolfSSL error: SSL_accept failed
> ```
This problem is on master as well and it's not specific to this PR (we don't have any DTLS1.3 configuration in the automatic test yet). I'm going to track down and fix this.

File: wolfssl/internal.h (hunk 8 above)
Comment (rizlik) [reply wolfssl/internal.h]: maybe `connectedSocket` or `connectedDatagramSocket`

**Comment by rizlik** [comment]
> > When I tested this locally, merged with master ([0459e83](https://github.com/wolfSSL/wolfssl/commit/0459e83a59d83ac7ebba3271a09ec26c52a56642)), the unit.test would fail:
> > ```
> > starting DTLSv1.3 suite - downgrade
> > trying server command line[3571]: SuiteTest -vd -7 2 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -2 -p 0 
> > trying client command line[3572]: SuiteTest -v 3 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -f -2 -p 51930 
> > SSL_accept error -326, record layer version error
> > wolfSSL error: SSL_accept failed
> > ```
> 
> This problem is on master as well and it's not specific to this PR (we don't have any DTLS1.3 configuration in the automatic test yet). I'm going to track down and fix this.

Indeed this was introduced by 060dfe1a693940fff31616570166be7f53b49ad8.

[#5315 ](https://github.com/wolfSSL/wolfssl/pull/5315) should fix this

File: src/internal.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (rizlik) [reply src/internal.c:8790]: Besides dev work logs are often asked for debugging non-local or not easily reproducible systems

**Review by rizlik — APPROVED** [review]

File: doc/dox_comments/header_files/ssl.h [hunk 10]
```diff
@@ -1631,9 +1631,40 @@ int  wolfSSL_set_fd (WOLFSSL* ssl, int fd);
...
     \sa wolfSSL_SetIOReadCtx
     \sa wolfSSL_SetIOWriteCtx
+    \sa wolfDTLS_SetChGoodCb
 */
 int wolfSSL_set_dtls_fd_connected(WOLFSSL* ssl, int fd)
 
+/*!
+    \ingroup Setup
+
+    \brief Allows setting a callback for DTLS client hello "good".
```
Comment (rizlik) [inline doc/dox_comments/header_files/ssl.h:1641]: it should be mentioned that if HRR Cookie is enabled, this CB is triggered only for CH with a valid cookie.

**Review by rizlik — CHANGES_REQUESTED** [review]

**Comment by julek-wolfssl** [comment]
@dgarske I can confirm that https://github.com/wolfSSL/wolfssl/pull/5315 solves @ejohnstown's issue.

**Comment by julek-wolfssl** [comment]
Rebased onto master and force pushed.

File: src/internal.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/internal.c:8790]: In that case I think its best to have all the info available ASAP.

File: wolfssl/internal.h (hunk 8 above)
Comment (julek-wolfssl) [reply wolfssl/internal.h]: I don't think that is a better name. Its too verbose and `connected` in the context of the `WOLFSSL_DTLS_CTX` object already only applies to the socket.

File: doc/dox_comments/header_files/ssl.h (hunk 10 above)
Comment (julek-wolfssl) [reply doc/dox_comments/header_files/ssl.h]: Added description.

**Review by rizlik — APPROVED** [review]

**Review by ejohnstown — APPROVED** [review]
The issue I was having was definitely outside of this PR.
